assembly/src/compile.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    /// Get the default [Options] for compiling a kernel module.
+    pub fn for_kernel() -> Self {
+        Self {
+            kind: ModuleKind::Kernel,
+            ..Default::default()
+        }
+    }
+}
+
+/// This error occurs when attempting to use a [Module] in a context which is invalid for that
+/// module.
+///
+/// For example, using an executable module as a library.
+#[derive(Debug, thiserror::Error, Diagnostic)]
+#[error("compilation failed: expected a {expected} module, but got a {actual} module")]
+#[diagnostic()]
+pub struct ModuleKindMismatchError {
+    expected: ModuleKind,
+    actual: ModuleKind,
+}
+
+// COMPILE TRAIT
+// ================================================================================================
+
+/// This trait is meant to be implemented by any type that can be compiled to a [Module],
+/// to allow methods which expect a [Module] to accept things like:
+///
+/// * A [Module] which was previously compiled or read from a [crate::Library].
+/// * A string representing the source code of a [Module].
+/// * A path to a file containing the source code of a [Module].
+/// * A vector of [Form]s comprising the contents of a [Module].
+pub trait Compile: Sized {
+    /// Compile (or convert) `self` into an executable [Module].
+    ///
+    /// See [Compile::compile_with_opts] for more details.
+    #[inline]
+    fn compile(self) -> Result<Box<Module>, Report> {
+        self.compile_with_opts(Options::default())
+    }
+
+    /// Compile (or convert) `self` into a [Module] using the provided `options`.
+    ///
+    /// Returns a [SyntaxError] if compilation fails due to a parsing or semantic analysis error,
+    /// or if the module provided is of the wrong kind (e.g. we expected a library module but got
+    /// an executable module).
+    ///
+    /// See the documentation for [Options] to see how compilation can be configured.
+    fn compile_with_opts(self, options: Options) -> Result<Box<Module>, Report>;
+}
+
+// COMPILE IMPLEMENTATIONS FOR MODULES
+// ------------------------------------------------------------------------------------------------
+
+impl Compile for Module {
+    #[inline(always)]
+    fn compile_with_opts(self, options: Options) -> Result<Box<Module>, Report> {
+        Box::new(self).compile_with_opts(options)
+    }
+}
+
+impl<'a> Compile for &'a Module {
+    #[inline(always)]
+    fn compile_with_opts(self, options: Options) -> Result<Box<Module>, Report> {
+        Box::new(self.clone()).compile_with_opts(options)
+    }
+}
+
+impl Compile for Box<Module> {
+    fn compile_with_opts(mut self, options: Options) -> Result<Box<Module>, Report> {
+        let actual = self.kind();
+        if actual == options.kind {
+            if let Some(path) = options.path {
+                self.set_path(path);
+            } else if let Some(ns) = options.namespace {
+                if !self.is_in_namespace(&ns) {
+                    self.set_namespace(ns);
+                }
```

Comment:
If I understood correctly, this changes just the namespace of a module while keeping the rest of the path intact. Under what circumstances is this useful? Maybe we should always just override module's path with the one provided via `options`?

Reply:
As mentioned in my other comment, the `namespace` option was made redundant by some changes I made after implementing this code, so it will go away, as will this.